test/kmerge-test.hs

```diff
@@ -425,11 +433,18 @@ loserTreeMerge xss = runST $ do
 -------------------------------------------------------------------------------}
 
 mutHeapMerge :: forall a. Ord a => [[a]] -> [a]
-mutHeapMerge xss = runST $ do
-    -- we reuse Heap.Entry structure here.
-    (heap, element) <- K.Heap.newMutableHeap [ Heap.Entry x xs | x:xs <- xss ]
-    go heap element
+mutHeapMerge xss = case [ Heap.Entry x xs | x :| xs <- cull xss ] of
+    [] -> []
+    e:es -> runST $ do
+      -- we reuse Heap.Entry structure here.
+      (heap, element) <- K.Heap.newMutableHeap $ e :| es
+      go heap $ Just element
   where
+    cull :: [[a]] -> [NonEmpty a]
+    cull = foldMap $ \case
+      [] -> []
+      x:xs -> [x :| xs]
```

Comment:
`cull` is not necessary, the previous code did the exact some thing. Doing a partial pattern match (e.g. `x:xs`) in a list comprehension skips the elements that didn't match.

Reply:
Ah yes, the `MonadFail`/`MonadPlus` condition of returning an empty list in the pattern match failure. That's exactly the functionality desired, and what was there previously.